File: backend/app/services/text_processor.py
"""文本处理服务 - 迁移自原项目 src/core/text_processor.py"""
import re
from typing import Optional, Tuple, List, Match


class TextProcessingError(Exception):
    """文本处理异常"""
    pass


class TextProcessor:
    """文本处理器类 - 从原项目迁移，保持业务逻辑不变"""

    @staticmethod
    def clean_text(text: str) -> str:
        """清理文本"""
        try:
            text = TextProcessor._remove_all_leading_spaces(text)
            text = TextProcessor._remove_empty_paragraphs(text)
            text = TextProcessor._insert_first_chapter(text)
            text = TextProcessor._separate_chapter_titles(text)
            text = TextProcessor._add_chapter_spacing(text)
            text = TextProcessor._add_paragraph_indent(text)
            return text
        except Exception as e:
            raise TextProcessingError(f"文本清理失败: {e}") from e

    @staticmethod
    def _remove_all_leading_spaces(text: str) -> str:
        return text.lstrip()

    @staticmethod
    def _remove_empty_paragraphs(text: str) -> str:
        return "\n".join([line for line in text.splitlines() if line.strip()])

    @staticmethod
    def _insert_first_chapter(text: str) -> str:
        return text

    @staticmethod
    def _separate_chapter_titles(text: str) -> str:
        lines = text.splitlines()
        result = []
        chapter_pattern = r'^\s*(序章|第[一二三四五六七八九十百千零0-9]+章)'
        for line in lines:
            match = re.match(chapter_pattern, line)
            if match:
                result.append(line.strip())
            elif line.strip():
                result.append(line.strip())
            elif line == '':
                result.append('')
        return "\n".join(result)

    @staticmethod
    def _add_chapter_spacing(text: str) -> str:
        lines = text.splitlines()
        result = []
        chapter_pattern = r'^\s*(序章|第[一二三四五六七八九十百千零0-9]+章)'
        for i, line in enumerate(lines):
            if re.match(chapter_pattern, line):
                if result and result[-1] != '':
                    result.append('')
                result.append(line.strip())
                if i < len(lines) - 1 and lines[i + 1].strip():
                    result.append('')
            elif line.strip():
                result.append(line.strip())
            elif line == '':
                result.append('')
        return "\n".join(result)

    @staticmethod
    def _add_paragraph_indent(text: str) -> str:
        lines = []
        chapter_pattern = r'^\s*(序章|第[一二三四五六七八九十百千零0-9]+章)'
        chinese_space = '\u3000'
        for line in text.splitlines():
            stripped_line = line.strip()
            if stripped_line:
                if re.match(chapter_pattern, stripped_line):
                    lines.append(stripped_line)
                else:
                    lines.append(chinese_space * 2 + stripped_line)
            else:
                lines.append('')
        return "\n".join(lines)

    @staticmethod
    def find_text(
        text: str,
        search_term: str,
        start_pos: int = 0,
        direction: str = 'next'
    ) -> Tuple[Optional[int], Optional[str]]:
        """查找文本"""
        try:
            if not search_term:
                return None, None
            pattern = re.compile(re.escape(search_term))
            if direction == 'next':
                match = pattern.search(text[start_pos:])
                if match:
                    return match.start() + start_pos, match.group()
            elif direction == 'previous':
                reverse_pos = start_pos - 1
                if reverse_pos >= 0:
                    match = pattern.search(text[:reverse_pos][::-1])
                    if match:
                        return reverse_pos - match.end(), match.group()
            return None, None
        except Exception as e:
            raise TextProcessingError(f"文本查找失败: {e}") from e

    @staticmethod
    def replace_all_word(lines: List[str], old_word: str, new_word: str) -> Tuple[List[str], int]:
        """替换所有"""
        try:
            result = []
            count = 0
            for line in lines:
                new_line = line.replace(old_word, new_word)
                count += line.count(old_word)
                result.append(new_line)
            return result, count
        except Exception as e:
            raise TextProcessingError(f"全部替换失败: {e}") from e

    @staticmethod
    def replace_except_in_quotes(
        lines: List[str],
        old_word: str,
        new_word: str
    ) -> Tuple[List[str], int]:
        """替换双引号外的指定单词"""
        try:
            count = 0

            def replace_match(match: Match) -> str:
                nonlocal count
                text = match.group(0)
                if (
                    text.startswith('\u201c') and text.endswith('\u201d')
                ) or (
                    text.startswith('\u201c') and text.endswith('\u201d')
                ):
                    return text
                replaced = text.replace(old_word, new_word)
                count += text.count(old_word)
                return replaced

            result = [re.sub(r'\u201c[^\u201d]*\u201d|"[^"]*"|[^\u201c\u201d"]+', replace_match, line) for line in lines]
            return result, count
        except Exception as e:
            raise TextProcessingError(f"双引号外替换失败: {e}") from e

File: frontend/lib/core/text_processor.dart
/// 文本处理核心模块
///
/// 提供文本清理、查找、替换等核心功能。
/// 功能等价于 Python 版 src/core/text_processor.py

class TextProcessor {
  static const String _chineseSpace = '\u3000';

  /// 章节标题正则：序章 或 第X章（中文数字+阿拉伯数字）
  static final RegExp _chapterPattern =
      RegExp(r'^\s*(序章|第[一二三四五六七八九十百千零0-9]+章)');

  /// 清理文本
  ///
  /// 执行以下操作：
  /// - 删除开头空格
  /// - 删除空白段落
  /// - 分离章节标题为独立段落
  /// - 章节间保留空行
  /// - 为段落开头添加两个中文空格缩进
  static String cleanText(String text) {
    text = _removeAllLeadingSpaces(text);
    text = _removeEmptyParagraphs(text);
    text = _insertFirstChapter(text);
    text = _separateChapterTitles(text);
    text = _addChapterSpacing(text);
    text = _addParagraphIndent(text);
    return text;
  }

  /// 删除开头空格
  static String _removeAllLeadingSpaces(String text) {
    // lstrip: remove leading whitespace
    return text.replaceFirst(RegExp(r'^\s+'), '');
  }

  /// 删除空白段落
  static String _removeEmptyParagraphs(String text) {
    return text.split('\n').where((line) => line.trim().isNotEmpty).join('\n');
  }

  /// 检测章节标题（保留用于向后兼容，不修改文本）
  static String _insertFirstChapter(String text) {
    return text;
  }

  /// 分离章节标题为独立段落
  static String _separateChapterTitles(String text) {
    final lines = text.split('\n');
    final result = <String>[];

    for (final line in lines) {
      if (_chapterPattern.hasMatch(line)) {
        result.add(line.trim());
      } else if (line.trim().isNotEmpty) {
        result.add(line.trim());
      } else if (line.isEmpty) {
        result.add('');
      }
    }

    return result.join('\n');
  }

  /// 在章节标题前后添加空行
  static String _addChapterSpacing(String text) {
    final lines = text.split('\n');
    final result = <String>[];

    for (var i = 0; i < lines.length; i++) {
      final line = lines[i];

      if (_chapterPattern.hasMatch(line)) {
        // 在章节标题前添加空行
        if (result.isNotEmpty && result.last.isNotEmpty) {
          result.add('');
        }
        result.add(line.trim());
        // 在章节标题后添加空行
        if (i < lines.length - 1 && lines[i + 1].trim().isNotEmpty) {
          result.add('');
        }
      } else if (line.trim().isNotEmpty) {
        result.add(line.trim());
      } else if (line.isEmpty) {
        result.add('');
      }
    }

    return result.join('\n');
  }

  /// 为段落开头添加两个中文空格缩进（章节标题除外）
  static String _addParagraphIndent(String text) {
    final lines = <String>[];

    for (final line in text.split('\n')) {
      final stripped = line.trim();
      if (stripped.isNotEmpty) {
        if (_chapterPattern.hasMatch(stripped)) {
          lines.add(stripped);
        } else {
          lines.add('$_chineseSpace$_chineseSpace$stripped');
        }
      } else {
        lines.add('');
      }
    }

    return lines.join('\n');
  }

  /// 查找文本中的目标词
  ///
  /// [direction]: 'next' 或 'previous'
  /// 返回 (位置, 匹配的文本)，未找到时返回 (null, null)
  static (int?, String?) findText(
    String text,
    String searchTerm, {
    int start_pos = 0,
    String direction = 'next',
  }) {
    if (searchTerm.isEmpty) return (null, null);

    try {
      final escaped = RegExp.escape(searchTerm);
      final pattern = RegExp(escaped);

      if (direction == 'next') {
        final match = pattern.search(text, start_pos);
        if (match != null) {
          return (match.start, match.group(0));
        }
      } else if (direction == 'previous') {
        final reversePos = start_pos - 1;
        if (reversePos >= 0) {
          final before = text.substring(0, reversePos);
          final reversed = before.split('').reversed.join();
          final match = pattern.search(reversed);
          if (match != null) {
            final pos = reversePos - match.end;
            return (pos, match.group(0));
          }
        }
      }

      return (null, null);
    } catch (e) {
      throw Exception('文本查找失败: $e');
    }
  }

  /// 替换所有行中的指定单词
  ///
  /// 返回 (替换后的行列表, 替换数量)
  static (List<String>, int) replaceAllWord(
    List<String> lines,
    String oldWord,
    String newWord,
  ) {
    final result = <String>[];
    var count = 0;

    for (final line in lines) {
      final occurrences = oldWord.allMatches(line).length;
      count += occurrences;
      result.add(line.replaceAll(oldWord, newWord));
    }

    return (result, count);
  }

  /// 替换双引号外的指定单词
  ///
  /// 中英文双引号内的内容不替换
  /// 返回 (替换后的行列表, 替换数量)
  static (List<String>, int) replaceExceptInQuotes(
    List<String> lines,
    String oldWord,
    String newWord,
  ) {
    var count = 0;
    final result = <String>[];

    // 匹配：中文双引号内容 | 英文双引号内容 | 非引号内容
    final quotePattern = RegExp(r'\u201c[^\u201d]*\u201d|"[^"]*"|[^\u201c\u201d"]+');

    for (final line in lines) {
      final replaced = line.replaceAllMapped(quotePattern, (match) {
        final text = match.group(0)!;
        // 检查是否为引号包裹的内容
        if ((text.startsWith('\u201c') && text.endsWith('\u201d')) ||
            (text.startsWith('"') && text.endsWith('"'))) {
          return text;
        }
        final occurrences = oldWord.allMatches(text).length;
        count += occurrences;
        return text.replaceAll(oldWord, newWord);
      });
      result.add(replaced);
    }

    return (result, count);
  }
}
